tools/cbmc/proofs/HTTP/global_state_HTTP.c

```diff
@@ -0,0 +1,201 @@
+    pConnInfo->pCaCert = malloc(sizeof(uint32_t));
+    pConnInfo->pClientCert = malloc(sizeof(uint32_t));
+    pConnInfo->pPrivateKey = malloc(sizeof(uint32_t));
+    pConnInfo->userBuffer.pBuffer = safeMalloc(sizeof(_connHandle_t));
+  }
+  return pConnInfo;
+}
+
+/* Creates a Connection Handle and assigns memory accordingly. */
+IotHttpsConnectionHandle_t newIotConnectionHandle () {
+  IotHttpsConnectionHandle_t pConnectionHandle = safeMalloc(sizeof(_connHandle_t));
+  if(pConnectionHandle) {
+    pConnectionHandle->pNetworkConnection = safeMalloc(sizeof(_connHandle_t));
+    pConnectionHandle->pNetworkInterface = newNetworkInterface();
+    pConnectionHandle->reqQ.pPrevious = safeMalloc(sizeof(IotLink_t));
+    pConnectionHandle->reqQ.pNext = safeMalloc(sizeof(IotLink_t));
+    pConnectionHandle->taskPoolJob = malloc(sizeof(struct _taskPoolJob));
+  }
+  return pConnectionHandle;
+}
+
+/* Creates a Response Handle and assigns memory accordingly. */
+IotHttpsResponseHandle_t newIotResponseHandle() {
+  IotHttpsResponseHandle_t pResponseHandle = safeMalloc(sizeof(_resHandle_t));
+  if(pResponseHandle) {
+    pResponseHandle->pBody = malloc(sizeof(uint32_t));
+    pResponseHandle->pHttpsConnection = newIotConnectionHandle();
+    pResponseHandle->pHttpsRequest = safeMalloc(sizeof(_reqHandle_t));
+    if(pResponseHandle->pHttpsRequest) {
+      pResponseHandle->pHttpsRequest->pHttpsResponse = pResponseHandle;
+    }
+    pResponseHandle->pHeaders = (uint8_t *)pResponseHandle;
+    pResponseHandle->pHeadersCur = pResponseHandle->pHeaders;
+    pResponseHandle->httpParserInfo.readHeaderParser.data = pResponseHandle; //???
+    pResponseHandle->httpParserInfo.parseFunc = http_parser_execute;
+    pResponseHandle->pReadHeaderValue = safeMalloc(sizeof(uint32_t));
+    pResponseHandle->pHeadersEnd = pResponseHandle->pHeaders + sizeof(_resHandle_t) ;
+    __CPROVER_assume(pResponseHandle->readHeaderValueLength >= 0 &&
+    pResponseHandle->readHeaderValueLength <= (pResponseHandle->pHeadersEnd - pResponseHandle->pHeaders));
+    pResponseHandle->pReadHeaderValue = malloc(pResponseHandle->readHeaderValueLength);
+  }
+  return pResponseHandle;
+}
+
+/* Creates a Request Handle and assigns memory accordingly. */
+IotHttpsRequestHandle_t newIotRequestHandle() {
+  IotHttpsRequestHandle_t requestHandle = safeMalloc(sizeof(_reqHandle_t));
+  if (requestHandle) {
+    requestHandle->pHttpsResponse = newIotResponseHandle();
+    requestHandle->pHttpsConnection = newIotConnectionHandle();
+    requestHandle->pBody = malloc(sizeof(uint32_t));
+    requestHandle->pHeaders = (uint8_t *)requestHandle;
+    requestHandle->pHeadersCur = requestHandle->pHeaders;
+    requestHandle->pHeadersEnd = requestHandle->pHeaders  + sizeof(_resHandle_t) ;
+    requestHandle->pConnInfo = newConnectionInfo();
+  }
+  return requestHandle;
+}
+
+/* Creates a Request Info and assigns memory accordingly. */
+IotHttpsRequestInfo_t * newIotRequestInfo() {
```

Comment:
I check for a NULL IotHttpsRequestInfo_t.pHost, does this affect the proof?
Do we need:
`pReqInfo->pHost >= 0 && pReqInfo->pHost <= IOT_HTTPS_MAX_HOST_NAME_LENGTH + 1`
?

Reply:
Allocated memory for this.